src/testdrive/src/action/kafka/ingest.rs

```diff
@@ -130,145 +179,107 @@ impl Action for IngestAction {
+                        .to_string(),
+                )
+            } else {
+                (None, row.clone())
+            };
+
+            fn decode_into_buf(
+                encoder: &Encoder,
+                row: &str,
+                buf: &mut Vec<u8>,
+            ) -> Result<(), String> {
+                match encoder {
+                    Encoder::Avro { schema, schema_id } => {
+                        let val = crate::format::avro::json_to_avro(
+                            &serde_json::from_str(row)
                                 .map_err(|e| format!("parsing avro datum: {}", e.to_string()))?,
-                            key_schema.top_node(),
+                            schema.top_node(),
                         )?;
-                        key_buf.write_u8(0).unwrap();
-                        key_buf.write_i32::<NetworkEndian>(*key_schema_id).unwrap();
-                        key_buf.extend(
-                            avro::to_avro_datum(&key_schema, key).map_err(|e| e.to_string())?,
-                        );
+                        // The first byte is a magic byte (0) that indicates the Confluent
+                        // serialization format version, and the next four bytes are a
+                        // 32-bit schema ID.
+                        //
+                        // https://docs.confluent.io/current/schema-registry/docs/serializer-formatter.html#wire-format
+                        buf.write_u8(0).unwrap();
+                        buf.write_i32::<NetworkEndian>(*schema_id).unwrap();
+                        buf.extend(avro::to_avro_datum(&schema, val).map_err(|e| e.to_string())?);
+                    }
+                    Encoder::Proto { parser, validator } => {
+                        let msg = parser(row)
+                            .map_err(|e| format!("converting row to type {} -> {}", row, e))?;
+                        *buf = msg
+                            .write_to_bytes()
+                            .map_err(|e| format!("writing protobuf message for {}: {}", row, e))?;
+                        // There are a variety of `write_*` methods on `Message` that don't
+                        // seem to automatically do the right thing. This should always
+                        // succeed, otherwise there is no chance for the server.
+                        let _parsed = validator(&buf).map_err(|e| {
+                            format!("error validating proto row={}\nerror={}", row, e)
+                        })?;
+                    }
+                    Encoder::Bytes => {
+                        *buf = row.as_bytes().to_vec();
+                    }
+                    Encoder::Text => {
+                        let json_string: serde_json::Value = serde_json::from_str(row)
+                            .map_err(|e| format!("parsing json string datum: {}", e.to_string()))?;
+                        *buf = json_string.as_str().unwrap().as_bytes().to_vec();
```

Comment:
As mentioned above, I don't understand this at all. Can't it just be `Encoder::Bytes`?